src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -329,14 +433,36 @@ private abstract class ReadOperation : AsyncOperation, IThreadPoolWorkItem
         {
             public ReadOperation(SocketAsyncContext context) : base(context) { }
 
-            void IThreadPoolWorkItem.Execute() => AssociatedContext.ProcessAsyncReadOperation(this);
+            void IThreadPoolWorkItem.Execute()
+            {
+                if (IsCompleted)
```

Comment:
for batched operations, we dispatch them to thread pool to more or less only run continuation on thread pool thread